src/server/pps/cmds/cmds.go

```diff
@@ -1071,191 +1075,391 @@ you can increase the amount of memory used for the bloom filters with the
 	); err != nil {
-		err = errors.Wrapf(err, "could not push docker image")
-		return "", err
+		return errors.Wrapf(err, "could not push docker image")
+	}
+
+	request.Transform.Image = destImage
+	return nil
+}
+
+// TODO: if transactions ever add support for pipeline creation, use them here
+// to create everything atomically
+func buildHelper(pc *pachdclient.APIClient, request *ppsclient.CreatePipelineRequest, pipelineParentPath string, update bool) error {
+	buildPath := request.Transform.Build.Path
+	if buildPath == "" {
+		buildPath = "."
+	}
+	if !filepath.IsAbs(buildPath) {
+		buildPath = filepath.Join(pipelineParentPath, buildPath)
+	}
+	if _, err := os.Stat(buildPath); err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("build path %q does not exist", buildPath)
+		}
+		return errors.Wrapf(err, "could not stat build path %q", buildPath)
+	}
+
+	buildPipelineName := fmt.Sprintf("%s_build", request.Pipeline.Name)
+
+	image := request.Transform.Build.Image
+	if image == "" {
+		pachctlVersion := version.PrettyPrintVersion(version.Version)
+		image = fmt.Sprintf("pachyderm/%s-build:%s", request.Transform.Build.Language, pachctlVersion)
+	}
+	if request.Transform.Image == "" {
+		request.Transform.Image = image
+	}
+
+	// utility function for creating an input used as part of a build step
+	createBuildPipelineInput := func(name string) *ppsclient.Input {
+		return &ppsclient.Input{
+			Pfs: &ppsclient.PFSInput{
+				Name:   name,
+				Glob:   "/",
+				Repo:   buildPipelineName,
+				Branch: name,
+			},
+		}
+	}
+
+	// create the source repo
+	if err := pc.UpdateRepo(buildPipelineName); err != nil {
+		return errors.Wrapf(err, "failed to create repo for build step-enabled pipeline")
+	}
+
+	// create the build pipeline
+	if err := pc.CreatePipeline(
+		buildPipelineName,
+		image,
+		[]string{"sh", "./build.sh"},
+		[]string{},
+		&ppsclient.ParallelismSpec{Constant: 1},
+		createBuildPipelineInput("source"),
+		"build",
+		update,
+	); err != nil {
+		return errors.Wrapf(err, "failed to create build pipeline for build step-enabled pipeline")
+	}
+
+	// retrieve ignores (if any)
+	ignores := []*glob.Glob{}
+	ignorePath := filepath.Join(buildPath, ".pachignore")
+	if _, err := os.Stat(ignorePath); err == nil {
+		f, err := os.Open(ignorePath)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read build step ignore file %q", ignorePath)
+		}
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			line := scanner.Text()
+			g, err := glob.Compile(line)
+			if err != nil {
+				return errors.Wrapf(err, "build step ignore file %q: failed to compile glob %q", ignorePath, line)
+			}
+			ignores = append(ignores, g)
+		}
+	}
+
+	// insert the source code
+	commit, err := pc.StartCommit(buildPipelineName, "source")
+	if err != nil {
+		return errors.Wrapf(err, "failed to start commit for source code in build step-enabled pipeline")
+	}
+	if update {
+		if err = pc.DeleteFile(buildPipelineName, commit.ID, "/"); err != nil {
```

Comment:
You can do this `DeleteFile` as an operation on the `pfc` which will make the entire commit atomic and not require calling `StartCommit` and `FinishCommit`.

Reply:
Neat!